slides/k8s/statefulsets.md

```diff
@@ -0,0 +1,402 @@
+
+- It is necessary to define a Storage Class to use [dynamic provisioning](https://kubernetes.io/docs/concepts/storage/dynamic-provisioning/)
+
+- Conversely, it is not necessary to define one if you will create volumes manually
+
+  (we will see dynamic provisioning in action later)
+
+---
+
+## Defining a Persistent Volume Claim
+
+Here is a minimal PVC:
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+   name: my-claim
+spec:
+   accessModes:
+     - ReadWriteOnce
+   resources:
+     requests:
+       storage: 1Gi
+```
+
+---
+
+## Using a Persistent Volume Claim
+
+Here is the same definition as earlier, but using a PVC:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-using-a-claim
+spec:
+  containers:
+  - image: ...
+    name: container-using-a-claim
+    volumeMounts:
+    - mountPath: /my-ebs
+      name: my-volume
+  volumes:
+  - name: my-volume
+    persistentVolumeClaim:
+      claimName: my-claim
+```
+
+---
+
+## Persistent Volume Claims and Stateful sets
+
+- The pods in a stateful set can define a `volumeClaimTemplate`
+
+- A `volumeClaimTemplate` will dynamically create one Persistent Volume Claim per pod
+
+- Each pod will therefore have its own volume
+
+- These volumes are numbered (like the pods)
+
+- When updating the stateful set (e.g. image upgrade), each pod keeps its volume
+
+- When pods get rescheduled (e.g. node failure), they keep their volume
+
+  (this requires a storage system that is not node-local)
+
+- These volumes are not automatically deleted
+
+  (when the stateful set is scaled down or deleted)
+
+---
+
+## Stateful set recap
+
+- A Stateful sets manages a number of identical pods
+
+  (like a Deployment)
+
+- These pods are numbered, and started/upgraded/stopped in a specific order
+
+- These pods are aware of their number
+
+  (e.g., #0 can decide to be the primary, and #1 can be secondary)
+
+- These pods can find the IP addresses of the other pods in the set
+
+  (through a *headless service*)
+
+- These pods can each have their own persistent storage
+
+  (Deployments cannot do that)
+
+---
+
+## Stateful sets in action
+
+- We are going to deploy a Consul cluster with 3 nodes
+
+- Consul is a highly-available key/value store
+
+  (like etcd or Zookeeper)
+
+- One easy way to bootstrap a cluster is to tell each node:
+
+  - the addresses of other nodes
+
+  - how many nodes are expected (to know when quorum is reached)
+
+---
+
+## Bootstrapping a Consul cluster
+
+*After reading the Consul documentation carefully (and/or asking around),
+we figure out the minimal command-line to run our Consul cluster.*
+
+```
+consul agent -data=dir=/consul/data -client=0.0.0.0 -server -ui \
+       -bootstrap-expect=3 \
+       -retry-join=`X.X.X.X` \
+       -retry-join=`Y.Y.Y.Y`
+```
+
+- We need to replace X.X.X.X and Y.Y.Y.Y with the addresses of other nodes
+
+- We can specify DNS names, but then they have to be FQDN
+
+- It's OK to specify ourselves in the list as well
```

Comment:
It's unclear to me what "ourselves" means here.

Reply:
Excellent point, I'll write "- It's OK for a pod to include itself in the list as well" instead.